File: mozilla/module_parametric.js
////////////////////////////////////////////////////////////////////////////////
// Parametric UV Shapes
////////////////////////////////////////////////////////////////////////////////

function planePos(u, v) {
  return [u - 0.5, 0, 0.5 - v];
}

function planeNor(u, v) {
  return [0, 1, 0];
}

function sphere(u, v) {
  const au = (2 * Math.PI) * u;
  const av = (1 * Math.PI) * v;
  const s = Math.sin(av);
  return [s * Math.cos(au), Math.cos(av), s * Math.sin(au)];
}

function torusPos(major, minor, u, v) {
  return function(u, v) {
    const au = (2 * Math.PI) * u;
    const av = (2 * Math.PI) * v;
    //x = (R + rcos(v))cos(u), y = (R + rcos(v))sin(u), and z = rsin(v). 
    return [
      (major + minor * Math.cos(av)) * Math.cos(au),
      minor * Math.sin(av),
      (major + minor * Math.cos(av)) * Math.sin(au),
    ];
  }
}

function torusNor(u, v) {
  return torusPos(0, 1)(u, v);
}

function unitUV(u, v) {
  return [u, v];
}

function getParametricPlane() {
  return { Pos : planePos, Nor : planeNor, UV0 : unitUV }
}

function getParametricSphere() {
  return { Pos : sphere, Nor : sphere, UV0 : unitUV }
}

function getParametricTorus(major, minor) {
  return { Pos : torusPos(major, minor), Nor : torusNor, UV0 : unitUV }
}

File: util/module_parametric.py
import math
from collections import namedtuple

###############################################################################
# Parametrics
###############################################################################

# A XZ plane
def planePos(u: float, v: float) -> tuple[float, float, float]:
   return (u - 0.5, 0, 0.5 - v)

def planeNor(u: float, v: float) -> tuple[float, float, float]:
   return [0, 1, 0]

# A sphere
def spherePos(u: float, v: float) -> tuple[float, float, float]:
  au = (2 * math.pi) * u
  av = (1 * math.pi) * v
  s = math.sin(av)
  return (s * math.cos(au), math.cos(av), s * math.sin(au))

def torusPos(major: float, minor: float):
   def function(u: float, v: float) -> tuple[float, float, float]:
      au = (2 * math.pi) * u
      av = (2 * math.pi) * v
      #x = (R + rcos(v))cos(u), y = (R + rcos(v))sin(u), and z = rsin(v). 
      return ((major + minor * math.cos(av)) * math.cos(au), minor * math.sin(av), (major + minor * math.cos(av)) * math.sin(au))
   return function

def unitUV(u: float, v: float) -> tuple[float, float]:
  return [u, v]

parametricClass = namedtuple("Parametric", ["Pos", "Nor", "ST0"])

def getParametricPlane() -> tuple:
   return parametricClass(Pos = planePos, Nor = planeNor, ST0 = unitUV)

def getParametricSphere() -> tuple:
   return parametricClass(Pos = spherePos, Nor = spherePos, ST0 = unitUV)

def getParametricTorus(major, minor) -> tuple:
   return parametricClass(Pos = torusPos(major, minor), Nor = torusPos(0, 1), ST0 = unitUV)